File: Python/Math/43. Multiply Strings.py
class Solution:
    def multiply(self, num1, num2):
        ans = [0 for _ in range(len(num1) + len(num2))]

        for i in range(len(num1) - 1, -1, -1):
            carry = 0
            for j in range(len(num2) - 1, -1, -1):
                tmp = ans[i + j + 1] - 0 + (ord(num1[i]) - ord('0')) * (ord(num2[j]) - ord('0')) + carry
                ans[i + j + 1] = tmp % 10
                carry = tmp//10
            ans[i] += carry

        pos = 0

        for i in range(len(ans)):
            if i == len(ans) - 1 and ans[i] == 0:
                return '0'
            if ans[i] == 0:
                pos += 1
            else:
                break
        return ''.join([str(i) for i in ans[pos:]])        

File: C++/Math/43. Multiply Strings.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        vector<int> ans(num1.size()+num2.size(), 0);
        for (int i = num1.size() - 1; i >= 0; i--) {
            int carry = 0;
            for (int j = num2.size() - 1; j >= 0; j--) {
                int prod = ans[i+j+1] + (num1[i] - '0') * (num2[j] - '0') + carry;
                ans[i+j+1] = prod % 10;
                carry = prod / 10;
            }
            ans[i] += carry;
        }
        int pos = 0;
        for (int i = 0; i < ans.size(); i++) {
            if (i == ans.size() - 1 && ans[i] == 0) {
                return "0";
            }
            else if (ans[i] == 0) {
                pos++;
            }
            else {
                break;
            }
        }
        string res = "";
        for (int i = pos; i < ans.size(); i++) {
            res += to_string(ans[i]);
        }
        return res;
    }
};